packages/devtools_app/lib/src/flutter/table.dart

```diff
@@ -0,0 +1,395 @@
+                    node: node,
+                  ),
+              ]),
+            ),
+          ),
+        ]),
+      ),
+    );
+  }
+}
+
+class _TreeNodeWidget<T extends TreeNode<T>> extends StatefulWidget {
+  const _TreeNodeWidget({
+    Key key,
+    @required this.node,
+    @required this.tableState,
+  }) : super(key: key);
+
+  final _TreeTableState<T> tableState;
+  final T node;
+
+  @override
+  _TreeNodeState createState() => _TreeNodeState<T>();
+}
+
+class _TreeNodeState<T extends TreeNode<T>> extends State<_TreeNodeWidget<T>>
+    with TickerProviderStateMixin {
+  // Animation controllers for bringing each node into the list,
+  // animating the size from 0 to the appropriate height.
+  AnimationController showController;
+  Animation<double> showAnimation;
+
+  // Animation controllers for animating the expand/collapse icon.
+  AnimationController expandController;
+  Animation<double> expandAnimation;
+
+  /// Whether or not this widget has been hidden.
+  ///
+  /// This is cached based on [TreeNode.shouldShow], which is
+  /// O([TreeNode.level]) to compute.
+  bool show;
+
+  // Convenience accessors for relevant fields from [TreeTable].
+  TreeColumnData<T> get treeColumn => widget.tableState.widget.treeColumn;
+  List<ColumnData<T>> get columns => widget.tableState.widget.columns;
+  String id(T data) =>
+      data == null ? null : widget.tableState.widget.keyFactory(data);
+  void onListUpdated() => widget.tableState.refreshTree();
+  List<double> get columnWidths => widget.tableState.columnWidths;
+
+  @override
+  void initState() {
+    super.initState();
+    showController = AnimationController(
+      vsync: this,
+      duration: const Duration(milliseconds: 200),
+    );
+    expandController = AnimationController(
+      vsync: this,
+      duration: const Duration(milliseconds: 200),
+    );
+    showAnimation = CurvedAnimation(
+      curve: Curves.easeInOutCubic,
+      parent: showController,
+    );
+    expandAnimation = Tween<double>(begin: 0.0, end: 0.5).animate(
+      CurvedAnimation(curve: Curves.easeInOutCubic, parent: expandController),
+    );
+    show = widget.node?.shouldShow() ?? true;
+    if (show) {
+      showController.forward();
+    }
+    if (widget.node?.isExpanded ?? false) {
+      expandController.forward();
+    }
+  }
+
+  @override
+  void dispose() {
+    showController.dispose();
+    expandController.dispose();
+    super.dispose();
+  }
+
+  void _toggleExpanded() {
+    _setExpanded(!widget.node.isExpanded);
+  }
+
+  void _setExpanded(bool isExpanded) {
+    setState(() {
+      if (isExpanded) {
+        widget.node.expand();
+        expandController.forward();
+      } else {
+        widget.node.collapse();
+        expandController.reverse();
+      }
+      onListUpdated();
+    });
+  }
+
+  @override
+  void didUpdateWidget(Widget oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    show = widget.node?.shouldShow() ?? true;
```

Comment:
I'm a bit suspicious of these animations doing the right thing in the didUpdateWidget case.
I think we are having nodes that end up with different state objects in problematic ways where animating between the different states won't really make sense. Happy to be wrong if there is something that will keep the same note assigned to the same state object for the typical cases.

Reply:
Right now, there is no view recycling going on.

If we do switch to a item builder pattern, then we may need to move the key up and apply it directly to the TreeNode widget, but keys are used for preserving the tree under rebuilds.